torch/csrc/autograd/profiler_utils.cpp

```diff
@@ -0,0 +1,201 @@
+      ss << "Failed to save extra arguments for flops compuation of op "
+         << op_name.str()
+         << ", min size: " << min_size
+         << ", actual size: " << inputs.size();
+      TORCH_WARN(ss.str());
+      return false;
+  }
+  for (auto index : should_be_tensor) {
+    if (!inputs[index].isTensor()) {
+      ss << "Failed to save extra arguments for flops compuation of op "
+         << op_name.str()
+         << ", input[" << index
+         << "] must be a tensor.";
+      TORCH_WARN(ss.str());
+      return false;
+    }
+  }
+  return true;
+}
+
+std::unordered_map<std::string, c10::IValue> saveExtraArgs(const at::RecordFunction& fn) {
+    // foe specific types of fn, return the saved extra args for computing flops
+    std::unordered_map<std::string, c10::IValue> map;
+    std::vector<c10::IValue> inputs = fn.inputs();
+
+    if (inputs.empty()) {
+      // Input shape is unavailable, return empty map
+      return map;
+    }
+
+    if (fn.name() == at::StringView("aten::conv2d")) {
+      std::vector<int> tensors{0, 1};
+      bool check = validateInput(fn.name(), CONV2D_GROUPS + 1, inputs, tensors);
+      if (!check) {
+        return map;
+      }
+
+      at::Tensor input = inputs[0].toTensor();
+      at::Tensor weight = inputs[1].toTensor();
+      if (weight.sizes().size() != 4) {
+        TORCH_WARN("Failed to compute flops for op aten::conv2d because it requires a 4D kernel tensor.");
+        return map;
+      }
+      map["input_size"] = at::IValue(input.sizes());
+      map["weight_size"] = at::IValue(weight.sizes());
+      map["stride"] = inputs[CONV2D_STRIDE];
+      map["padding"] = inputs[CONV2D_PADDING];
+      map["dilation"] = inputs[CONV2D_DILATION];
+      map["groups"] = inputs[CONV2D_GROUPS];
+    } else if (fn.name() == at::StringView("aten::mm")) {
+      std::vector<int> tensors{0, 1};
+      bool check = validateInput(fn.name(), 2, inputs, tensors);
+      if (!check) {
+        return map;
+      }
+
+      at::Tensor left = inputs[0].toTensor();
+      at::Tensor right = inputs[1].toTensor();
+      map["mat1_size"] = at::IValue(left.sizes());
+      map["mat2_size"] = at::IValue(right.sizes());
+    } else if (fn.name() == at::StringView("aten::mul.Tensor")) {
+      std::vector<int> tensors{0};
+      bool check = validateInput(fn.name(), 1, inputs, tensors);
+      if (!check) {
+        return map;
+      }
+
+      at::Tensor mat = inputs[0].toTensor();
+      map["mat_size"] = at::IValue(mat.sizes());
+    } else if (fn.name() == at::StringView("aten::add.Tensor")) {
+      std::vector<int> tensors{0};
+      bool check = validateInput(fn.name(), 1, inputs, tensors);
+      if (!check) {
+        return map;
+      }
+
+      at::Tensor mat = inputs[0].toTensor();
+      map["mat_size"] = at::IValue(mat.sizes());
+    }
+
+    return map;
+}
+
+uint64_t computeFlops(const at::StringView &op_name, const std::unordered_map<std::string, c10::IValue> &extra_args) {
```

Comment:
for future: if you plan to add a ton of overrides, it might be worth to use some of the boxing/unboxing machinery here. Then it'd be cool that we can write flops inference function just as a regular C++ function with the same signature but where Tensors are instead TensorMeta without the actual data. It shouldn't block this diff, but would be good to consider. cc @ezyang

Reply:
Totally agree. It will be cool to add a TensorMeta layer that contains all data needed for flops inference. It makes contribution to flops inference function much easier. What do you think of writing flops inference functions in C++ vs. in Python?

Reply:
I have to say, I am a little surprised that you guys didn't decide to put the FLOP calculations in Python. But yeah, if this is living in C++, we can consider a tighter integration with how people write operators, so that you can (1) reuse our existing operator extensibility machinery to let people writing custom ops easily define FLOP formulas for their stuff and (2) make people more likely to write in FLOPs for new operators they write.

Reply:
we will consider that in the follow up, let's continue the discussion on how to make a tighter integration of flops formula and the op